### PR TITLE
ASoC: SOF: fix snd_sof_fw_parse_ext_data

### PR DESCRIPTION
**Pull request by ktrzcinx**
There was error during parsing more than one ext_data from mailbox,
because of invalid data offset handling.
Moreover added warning log when found some unknown FW boot ext header.

Signed-off-by: Karol Trzcinski <karolx.trzcinski@linux.intel.com>

This PR is needed for PR https://github.com/thesofproject/linux/pull/1526/commits

### CODE REVIEW

File: sound/soc/sof/loader.c [hunk 1]
```diff
@@ -50,8 +50,7 @@ int snd_sof_fw_parse_ext_data(struct snd_sof_dev *sdev, u32 bar, u32 offset)
 
 	while (ext_hdr->hdr.cmd == SOF_IPC_FW_READY) {
 		/* read in ext structure */
```
Comment (dbaluta) [inline sound/soc/sof/loader.c:52]: @ktrzcinx there are two different issues here. Please make two separate patches.

File: sound/soc/sof/loader.c [hunk 2]
```diff
@@ -50,8 +50,7 @@ int snd_sof_fw_parse_ext_data(struct snd_sof_dev *sdev, u32 bar, u32 offset)
...
-		offset += sizeof(*ext_hdr);
-		snd_sof_dsp_block_read(sdev, bar, offset,
+		snd_sof_dsp_block_read(sdev, bar, offset + sizeof(*ext_hdr),
```
Comment (ranj063) [inline sound/soc/sof/loader.c:53]: @ktrzcinx is the issue because ext_hdr->hdr.size already includes the header size in it?

File: sound/soc/sof/loader.c [hunk 3]
```diff
@@ -61,11 +60,13 @@ int snd_sof_fw_parse_ext_data(struct snd_sof_dev *sdev, u32 bar, u32 offset)
 		/* process structure data */
 		switch (ext_hdr->type) {
 		case SOF_IPC_EXT_DMA_BUFFER:
+			ret = 0;
```
Comment (ranj063) [inline sound/soc/sof/loader.c:63]: can we reset ret to 0 outside the switch?

File: sound/soc/sof/loader.c [hunk 4]
```diff
@@ -61,11 +60,13 @@ int snd_sof_fw_parse_ext_data(struct snd_sof_dev *sdev, u32 bar, u32 offset)
...
 			break;
 		case SOF_IPC_EXT_WINDOW:
 			ret = get_ext_windows(sdev, ext_hdr);
 			break;
 		default:
+			ret = 0;
```
Comment (plbossart) [inline sound/soc/sof/loader.c:71]: please don't submit two PRs which touch the same lines of code. It will lead to conflicts.
Also why is the point of changing the return value? get _ext_windows can only return 0, -EINVAL and -ENOMEM and the last two are trapped. doing those additional inits does not bring any added value. If you want to be explicit and not force the reader to reconstruct the flow, move this outside of each switch case so that the value is reset for each new header.

File: sound/soc/sof/loader.c (hunk 2 above)
Comment (ktrzcinx) [reply sound/soc/sof/loader.c:53]: Yes

Comment (ranj063) [reply sound/soc/sof/loader.c:53]: OK cool. Can you please update the commit message to clarify this?

File: sound/soc/sof/loader.c (hunk 4 above)
Comment (ktrzcinx) [reply sound/soc/sof/loader.c:71]: > please don't submit two PRs which touch the same lines of code

I know, but I do as @dbaluta wish in this PR

> move this outside of each switch case

> can we reset ret to 0 outside the switch?

Do you really like such a solution? I prefer my way or make something like
```
switch(...) {
  case 1:
  case 2:
    [..]
}
```

Because it is the most explicit solution, it binds together case=>action in straightforward way, so reader can't be confused.

File: sound/soc/sof/loader.c [hunk 5]
```diff
@@ -61,11 +60,13 @@ int snd_sof_fw_parse_ext_data(struct snd_sof_dev *sdev, u32 bar, u32 offset)
...
 			break;
 		}
```
Comment (plbossart) [inline sound/soc/sof/loader.c:73]: please fix the conflict and the commit message (using 'there is' is too vague, use more active wording)

An error occurs during parsing more than one ext_data from the mailbox, because of invalid data offset handling. Fix by removing the incorrect duplicate increment of the offset.

The return value is also reset in the switch case. This does not change the behavior but improves readability - there is no longer a need to check what the return value of get_ext_windows is.

Comment (plbossart) [reply sound/soc/sof/loader.c:73]: note as @dbaluta suggested that the second part should be in a separate patch, it's not logically tied to the offset change.

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
@fredoh9 can you help check CI here? it's still pending-building but the status is ok, and the device tests are surprising. I'd like to restart a clean test, but SOFCI TEST doesn't seem to do anything?
